tests/qnn/test_keras.py

```diff
@@ -0,0 +1,425 @@
+"""
+import os
+import tempfile
+
+import numpy as np
+import pytest
+
+import pennylane as qml
+from pennylane.qnn.keras import KerasLayer
+
+tf = pytest.importorskip("tensorflow", minversion="2")
+
+
+@pytest.fixture
+def get_circuit(n_qubits, output_dim, interface):
+    """Fixture for getting a sample quantum circuit with a controllable qubit number and output
+    dimension. Returns both the circuit and the shape of the weights."""
+
+    dev = qml.device("default.qubit", wires=n_qubits)
+    weight_shapes = {
+        "w1": (3, n_qubits, 3),
+        "w2": (1,),
+        "w3": 1,
+        "w4": [3],
+        "w5": (2, n_qubits, 3),
+        "w6": 3,
+        "w7": 0,
+    }
+
+    @qml.qnode(dev, interface=interface)
+    def circuit(inputs, w1, w2, w3, w4, w5, w6, w7):
+        """A circuit that embeds data using the AngleEmbedding and then performs a variety of
+        operations. The output is a PauliZ measurement on the first output_dim qubits. One set of
+        parameters, w5, are specified as non-trainable."""
+        qml.templates.AngleEmbedding(inputs, wires=list(range(n_qubits)))
+        qml.templates.StronglyEntanglingLayers(w1, wires=list(range(n_qubits)))
+        qml.RX(w2[0], wires=0 % n_qubits)
+        qml.RX(w3, wires=1 % n_qubits)
+        qml.Rot(*w4, wires=2 % n_qubits)
+        qml.templates.StronglyEntanglingLayers(w5, wires=list(range(n_qubits)))
+        qml.Rot(*w6, wires=3 % n_qubits)
+        qml.RX(w7, wires=4 % n_qubits)
+        return [qml.expval(qml.PauliZ(i)) for i in range(output_dim)]
+
+    return circuit, weight_shapes
+
+
+@pytest.mark.usefixtures("get_circuit")
+@pytest.fixture
+def model(get_circuit, n_qubits, output_dim):
+    """Fixture for creating a hybrid Keras model. The model is composed of KerasLayers sandwiched
+    between Dense layers."""
+    c, w = get_circuit
+    layer1 = KerasLayer(c, w, output_dim)
+    layer2 = KerasLayer(c, w, output_dim)
+
+    model = tf.keras.models.Sequential(
+        [
+            tf.keras.layers.Dense(n_qubits),
+            layer1,
+            tf.keras.layers.Dense(n_qubits),
+            layer2,
+            tf.keras.layers.Dense(output_dim),
+        ]
+    )
+
+    return model
+
+
+def indicies_up_to(n_max):
+    """Returns an iterator over the number of qubits and output dimension, up to value n_max.
+    The output dimension never exceeds the number of qubits."""
+    a, b = np.tril_indices(n_max)
+    return zip(*[a + 1, b + 1])
+
+
+@pytest.mark.usefixtures("get_circuit")
+class TestKerasLayer:
+    """Unit tests for the pennylane.qnn.keras.KerasLayer class."""
+
+    @pytest.mark.parametrize("interface", ["tf"])
+    @pytest.mark.parametrize("n_qubits, output_dim", indicies_up_to(1))
+    def test_bad_tf_version(self, get_circuit, output_dim, monkeypatch):
+        """Test if an ImportError is raised when instantiated with an incorrect version of
+        TensorFlow"""
+        c, w = get_circuit
+        with monkeypatch.context() as m:
```

Comment:
Nice, didn't know that one!

Reply:
I understand, this is truly a unit test if you engineer the function to fulfil your test case. 

But on the other hand, wouldn't it be good to actually test the *user behaviour* you want to catch, for example by loading another tf version here?

Reply:
This would be also true for everything below...can you really be sure that the wrong usage of the class leads to these things simulated by monkeypatch?

Reply:
Yep, that's why we generally have unit tests and integration tests. I think @trbromley has focused on the unit tests at the moment, since this is a new module

Reply:
Yes this one is definitely a fairly contrived unit test, although I couldn't think of a smarter way. For example, I tried monkeypatching the version of tensorflow to be a `0.0.0` but couldn't seem to get the test to notice this :thinking: